### ISSUE TITLE
Mandatory / required filter in requests

### ISSUE THREAD

**Issue by bastians** [mevdschee/php-crud-api#687]
Is it possible to mark a filter as required that the API throws an error when the filter is used but no parameter is given in the request?

**Comment by mevdschee** [comment]
Thank you for creating this issue.

The answer is "No" as this is not (yet) implemented. You may implement such behavior in a 'beforeHandler' of the customization middleware. You could also make a new middleware for this functionality. 

Maybe you can explain a bit more about your use-case and motivation? It could even be a candidate for core functionality if it seems essential.

**Comment by bastians** [comment]
Wow, this answer was quick, thank you :-)

I have a date input field on my app and this input field is actually required. I can make this field required on the app side, of course, but somehow for me it feels more correct to have the validation also on backend side. Maybe this is just a personal feeling, but my approach is that the "app should be dump". :-)

**Comment by mevdschee** [comment]
Ah.. I understand.  I agree that validation should be done on the backend, but this is a different kind of validation, as you limit the way people ask for data, right? I would instead limit the size of the response, as that is a more universal problem. You can try the "pageLimits" middleware and see whether or not that fits your needs.

**Comment by bastians** [comment]
> ...as you limit the way people ask for data, right?

Yes, but limiting the response does not solve my problem. The input is from type "date" and based on the date, the user gets a result. But without the date, the request does not make sense and should fail because currently, it just shows all data in the result.

**Comment by mevdschee** [comment]
You can write a custom "beforeFilter" using the customization middleware and give an error when the date is not given as a filter parameter.

**Comment by bastians** [comment]
Ok, perfect, then I will try this :-) Thank you!

**Comment by dnmd** [comment]
I have the same requirement and initially used the straightforward implementation suggested in https://github.com/mevdschee/php-crud-api/issues/750

```
'customization.beforeHandler' => function ($operation, $tableName, $request, $environment) {
    $in = "GET /unknown_url";
    return RequestFactory::fromString($in);
},
```
However, a 400 error with a message like "Invalid query parameters" would be more appropriate. Is there a way to directly return this within the `beforeHandler`, or is it only possible in the `afterHandler`?

**Comment by mevdschee** [comment]
>Is there a way to directly return this within the beforeHandler, or is it only possible in the afterHandler?

No, you need to use the "environment" variable to pass the data from the "beforehandler" to the "afterhandler" and output the response there.

**Comment by dnmd** [comment]
The implementation is currently as follows:

```
'customization.beforeHandler' => function ($operation, $tableName, $request, $environment) {
    $environment->filterMissing = false;

    if ($tableName === 'table_name' && $operation === 'list' && empty($request->getQueryParams()['filter'])) {
        $environment->filterMissing = true;
    }
},
'customization.afterHandler' => function ($operation, $tableName, $response, $environment) {
    if ($environment->filterMissing) {
        return ResponseFactory::fromObject(400, ['code' => 1040, 'message' => 'Filter parameter is missing'], 0);
    }
},
```

I did notice that some headers are slightly different in this custom response e.g. `connection: close` instead of `Keep-Alive`, so I'll try to find a way to get this right as well.

**Comment by dnmd** [comment]
Regarding the specific question originally raised by the OP, I believe it can be considered resolved. However, I did notice some curious behavior in the `customization.afterHandler`, with regard to the `Connection` header as I previously mentioned:

`return ResponseFactory::fromStatus(404)` => `Connection: Keep-Alive`
`return ResponseFactory::fromStatus(401)` => `Connection: Keep-Alive`
`return ResponseFactory::fromStatus(400)` => `Connection: close`

Maybe this is normal behavior, as I am unfamiliar with the spec of the Psr implementation. Using `->withHeader('Connection', 'Keep-Alive')` only appends the value making it `Connection: Keep-Alive, close`. Any suggestion is welcome, I can raise the question as well on the https://github.com/Nyholm/psr7 if that is more appropriate.
